File: app/state/runtime_trace.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any
from uuid import uuid4

from pydantic import BaseModel, Field

from app.state.runtime_state import DecisionBoundaryResult, SafetyGateResult


def _utc_now() -> datetime:
    return datetime.now(timezone.utc)


def _generate_trace_id() -> str:
    return f"trace_{uuid4().hex[:12]}"


class RuntimeTrace(BaseModel):
    trace_id: str = Field(default_factory=_generate_trace_id)
    runtime_id: str
    step: int = 1
    input: str = ""
    modules_executed: list[str] = Field(default_factory=list)
    knowledge_used: list[str] = Field(default_factory=list)
    experience_used: list[str] = Field(default_factory=list)
    safety_gate_result: SafetyGateResult | None = None
    ddx_change: dict[str, Any] | None = None
    evidence_graph_change: dict[str, Any] | None = None
    decision_boundary_result: DecisionBoundaryResult | None = None
    output_summary: dict[str, Any] | None = None
    created_at: datetime = Field(default_factory=_utc_now)

    def record_module(self, module_name: str) -> None:
        if module_name not in self.modules_executed:
            self.modules_executed.append(module_name)

    def record_knowledge(self, asset: str) -> None:
        if asset not in self.knowledge_used:
            self.knowledge_used.append(asset)

    def record_experience(self, unit_id: str) -> None:
        if unit_id not in self.experience_used:
            self.experience_used.append(unit_id)

File: src/main/java/com/clinmind/runtime/state/RuntimeTrace.java
package com.clinmind.runtime.state;

import java.time.Instant;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class RuntimeTrace {

    private String traceId;
    private String runtimeId;
    private int step;
    private String input;
    private List<String> modulesExecuted;
    private List<String> knowledgeUsed;
    private List<String> experienceUsed;
    private SafetyGateResult safetyGateResult;
    private Map<String, Object> ddxChange;
    private Map<String, Object> evidenceGraphChange;
    private DecisionBoundaryResult decisionBoundaryResult;
    private Map<String, Object> outputSummary;
    private Instant createdAt;

    public RuntimeTrace() {
    }

    public static RuntimeTrace create(String runtimeId, int step, String input) {
        RuntimeTrace trace = new RuntimeTrace();
        trace.traceId = IdGenerator.traceId();
        trace.runtimeId = runtimeId;
        trace.step = step;
        trace.input = input == null ? "" : input;
        trace.modulesExecuted = new ArrayList<>();
        trace.knowledgeUsed = new ArrayList<>();
        trace.experienceUsed = new ArrayList<>();
        trace.createdAt = Instant.now();
        return trace;
    }

    public void recordModule(String moduleName) {
        if (moduleName != null && !modulesExecuted.contains(moduleName)) {
            modulesExecuted.add(moduleName);
        }
    }

    public void recordKnowledge(String asset) {
        if (asset != null && !knowledgeUsed.contains(asset)) {
            knowledgeUsed.add(asset);
        }
    }

    public void recordExperience(String unitId) {
        if (unitId != null && !experienceUsed.contains(unitId)) {
            experienceUsed.add(unitId);
        }
    }

    public String getTraceId() {
        return traceId;
    }

    public void setTraceId(String traceId) {
        this.traceId = traceId;
    }

    public String getRuntimeId() {
        return runtimeId;
    }

    public void setRuntimeId(String runtimeId) {
        this.runtimeId = runtimeId;
    }

    public int getStep() {
        return step;
    }

    public void setStep(int step) {
        this.step = step;
    }

    public String getInput() {
        return input;
    }

    public void setInput(String input) {
        this.input = input;
    }

    public List<String> getModulesExecuted() {
        return modulesExecuted;
    }

    public void setModulesExecuted(List<String> modulesExecuted) {
        this.modulesExecuted = modulesExecuted;
    }

    public List<String> getKnowledgeUsed() {
        return knowledgeUsed;
    }

    public void setKnowledgeUsed(List<String> knowledgeUsed) {
        this.knowledgeUsed = knowledgeUsed;
    }

    public List<String> getExperienceUsed() {
        return experienceUsed;
    }

    public void setExperienceUsed(List<String> experienceUsed) {
        this.experienceUsed = experienceUsed;
    }

    public SafetyGateResult getSafetyGateResult() {
        return safetyGateResult;
    }

    public void setSafetyGateResult(SafetyGateResult safetyGateResult) {
        this.safetyGateResult = safetyGateResult;
    }

    public Map<String, Object> getDdxChange() {
        return ddxChange;
    }

    public void setDdxChange(Map<String, Object> ddxChange) {
        this.ddxChange = ddxChange;
    }

    public Map<String, Object> getEvidenceGraphChange() {
        return evidenceGraphChange;
    }

    public void setEvidenceGraphChange(Map<String, Object> evidenceGraphChange) {
        this.evidenceGraphChange = evidenceGraphChange;
    }

    public DecisionBoundaryResult getDecisionBoundaryResult() {
        return decisionBoundaryResult;
    }

    public void setDecisionBoundaryResult(DecisionBoundaryResult decisionBoundaryResult) {
        this.decisionBoundaryResult = decisionBoundaryResult;
    }

    public Map<String, Object> getOutputSummary() {
        return outputSummary;
    }

    public void setOutputSummary(Map<String, Object> outputSummary) {
        this.outputSummary = outputSummary == null ? null : new LinkedHashMap<>(outputSummary);
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(Instant createdAt) {
        this.createdAt = createdAt;
    }
}
